tests/functional/test_import.py

```diff
@@ -47,14 +47,105 @@ def test_ctypes_CDLL_c(pyi_builder):
         """)
 
 
-# libnss_files.so should be available in most Unix/Linux systems using
-# glibc
-@skipif(not ctypes.util.find_library('nss_files'),
-        reason="required libnss_files.so missing")
-def test_ctypes_CDLL_find_library__nss_files(pyi_builder):
-    pyi_builder.test_source(
-        """
-        import ctypes, ctypes.util
-        lib = ctypes.CDLL(ctypes.util.find_library('nss_files'))
-        assert lib is not None
-        """)
+def skip_if_lib_missing(libname, text=None):
+    """
+    pytest decorator to evaluate the required shared lib.
+
+    :param libname: Name of the required library.
+    :param text: Text to put into the reason message
+                 (defaults to 'lib%s.so' % libname)
+
+    :return: pytest decorator with a reason.
+    """
+    soname = ctypes.util.find_library(libname)
+    if not text:
+        text = "lib%s.so" % libname
+    # Return pytest decorator.
+    return skipif(not (soname and ctypes.CDLL(soname)),
+                  reason="required %s missing" % text)
+
+
+_template_ctypes_CDLL_find_library = """
+    import ctypes, ctypes.util, sys, os
+    lib = ctypes.CDLL(ctypes.util.find_library(%(libname)r))
+    print(lib)
+    assert lib is not None and lib._name is not None
+    if getattr(sys, 'frozen', False):
+        soname = ctypes.util.find_library(%(libname)r)
+        print(soname)
+        libfile = os.path.join(sys._MEIPASS, soname)
+        print(libfile)
+        assert os.path.isfile(libfile), '%%s is missing' %% soname
+        print('>>> file found')
+    """
+
+_template_ctypes_test = """
+        print(lib)
+        assert lib is not None and lib._name is not None
+        import sys, os
+        if getattr(sys, 'frozen', False):
+            libfile = os.path.join(sys._MEIPASS, %(soname)r)
+            print(libfile)
+            assert os.path.isfile(libfile), '%(soname)s is missing'
+            print('>>> file found')
+    """
+
+# Ghostscript's libgs.so should be available in may Unix/Linux systems
+@skip_if_lib_missing('gs', 'libgs.so (Ghostscript)')
+def test_ctypes_CDLL_find_library__gs(pyi_builder):
+    libname = 'gs'
+    pyi_builder.test_source(_template_ctypes_CDLL_find_library % locals())
+
+
+#-- Generate test-cases for the different types of ctypes objects.
+
+libname = 'gs'
+reason = 'libgs.so (Ghostscript)'
+parameters = []
+ids = []
+for prefix in ('', 'ctypes.'):
+    for funcname in  ('CDLL', 'PyDLL', 'WinDLL', 'OleDLL', 'cdll.LoadLibrary'):
+        ids.append('%s_%s' % (prefix+funcname, libname))
+        params = (prefix+funcname, libname, reason, ids[-1])
+        if funcname in ("WinDLL", "OleDLL"):
+            # WinDLL, OleDLL only work on windows.
+            params = skipif_notwin(params)
```

Comment:
[Actually, there's something about this in the py.test documentation](https://pytest.org/latest/example/markers.html#marking-platform-specific-tests-with-pytest)

Adding a `runtest_setup` hook doesn't look like a big deal. They describe how to create custom marks that say "only runs on OS X" and similar, and then use the hook function to skip based on the custom marks instead of doing everything through `skipif`. This way, you can stack the `osx` mark with the `skipif` mark without worrying about two `skipif` marks replacing one another.